src/server.rs

```diff
@@ -1,89 +1,198 @@
 use serde::{Deserialize, Serialize};
-use std::net::SocketAddr;
-use std::sync::Mutex;
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex};
 
 #[derive(Clone, Debug, PartialEq)]
-pub struct ServerConfig {
-    pub key: SecretKey,
-    pub addr: SocketAddr,
-    pub is_first: bool,
-}
-
-pub struct Submission {
-    pub excess: SecretKey,
-    pub input_commit: Commitment,
-    pub onion: Onion,
+struct Submission {
+    excess: SecretKey,
+    output_commit: Commitment,
+    rangeproof: Option<RangeProof>,
+    input: Input,
+    fee: u64,
+    onion: Onion,
 }
 
 #[derive(Serialize, Deserialize)]
 pub struct SwapReq {
-    pub onion: Onion,
-    #[serde(with = "ser::vec_serde")]
-    pub msg: Vec<u8>,
+    onion: Onion,
     #[serde(with = "secp::comsig_serde")]
-    pub comsig: ComSignature,
+    comsig: ComSignature,
 }
 
 lazy_static! {
-    static ref SERVER_STATE: Mutex<Vec<Submission>> = Mutex::new(Vec::new());
+    static ref SERVER_STATE: Mutex<HashMap<Commitment, Submission>> = Mutex::new(HashMap::new());
 }
 
 #[rpc(server)]
 pub trait Server {
     #[rpc(name = "swap")]
     fn swap(&self, swap: SwapReq) -> Result<Value>;
 
-    // milestone 3:
+    // milestone 3: Used by mwixnet coinswap servers to communicate with each other
     // fn derive_outputs(&self, entries: Vec<Onion>) -> Result<Value>;
     // fn derive_kernel(&self, tx: Tx) -> Result<Value>;
 }
 
-pub struct ServerImpl {
-    server_key: SecretKey,
+#[derive(Clone)]
+struct ServerImpl {
+    server_config: ServerConfig,
+    stop_state: Arc<StopState>,
+    wallet: Arc<dyn Wallet>,
+    node: Arc<dyn GrinNode>,
 }
 
 impl ServerImpl {
-    pub fn new(server_key: SecretKey) -> Self {
-        ServerImpl { server_key }
+    fn new(server_config: ServerConfig, stop_state: Arc<StopState>, wallet: Arc<dyn Wallet>, node: Arc<dyn GrinNode>) -> Self {
+        ServerImpl { server_config, stop_state, wallet, node }
+    }
+
+    /// The fee base to use. For now, just using the default.
+    fn get_fee_base(&self) -> u64 {
+        DEFAULT_ACCEPT_FEE_BASE
+    }
+
+    /// Minimum fee to perform a swap.
+    /// Requires enough fee for the mwixnet server's kernel, 1 input and its output to swap.
+    fn get_minimum_swap_fee(&self) -> u64 {
+        TransactionBody::weight_by_iok(1, 1, 1) * self.get_fee_base()
+    }
+
+    /// Iterate through all saved submissions, filter out any inputs that are no longer spendable,
+    /// and assemble the coinswap transaction, posting the transaction to the configured node.
+    /// 
+    /// Currently only a single mix node is used. Milestone 3 will include support for multiple mix nodes.
+    fn execute_round(&self) -> crate::error::Result<()> {
+        let mut locked_state = SERVER_STATE.lock().unwrap();
+        let next_block_height = self.node.get_chain_height()? + 1;
+
+        let spendable : Vec<Submission> = locked_state
+            .values()
+            .into_iter()
+            .unique_by(|s| s.output_commit)

```

Comment:
I assume this resolves the [duplicate outputs](https://github.com/mimblewimble/mwixnet/issues/2) scenario. The next filter ensure the input is spendable and the filter after that ensures that the ending commitment isn't already present in the UTXO set. I think the duplicate input check might be missing for the case when someone makes two SwapReq for the same input

Reply:
Duplicate input checking is done as part of the [swap handler](https://github.com/mimblewimble/mwixnet/blob/bff59b08ac14ba4ea3617bc2add71ae9e3cccd96/src/server.rs#L173-L175) . Duplicate inputs can't exist at this point.